pkg/agent/watch.go

```diff
@@ -129,6 +51,41 @@ func startVMWatcher(
 		},
 		util.InitWatchModeDefer,
 		metav1.ListOptions{},
-		util.WatchHandlerFuncs[*vmapi.VirtualMachine]{},
+		util.WatchHandlerFuncs[*vmapi.VirtualMachine]{
+			AddFunc: func(vm *vmapi.VirtualMachine, preexisting bool) {
+				if vmIsOurResponsibility(vm, config, nodeName) {
+					podEvents <- makeVMEvent(vm, vmEventAdded)
+				}
+			},
+			UpdateFunc: func(oldVM, newVM *vmapi.VirtualMachine) {
+				oldIsOurs := vmIsOurResponsibility(oldVM, config, nodeName)
+				newIsOurs := vmIsOurResponsibility(newVM, config, nodeName)
+
+				if !oldIsOurs && newIsOurs {
+					podEvents <- makeVMEvent(newVM, vmEventAdded)
+				} else if oldIsOurs && !newIsOurs {
+					podEvents <- makeVMEvent(oldVM, vmEventDeleted)
+				}
+			},
+			DeleteFunc: func(vm *vmapi.VirtualMachine, maybeStale bool) {
+				if vmIsOurResponsibility(vm, config, nodeName) {
+					podEvents <- makeVMEvent(vm, vmEventDeleted)
+				}
+			},
+		},
 	)
 }
+
+func makeVMEvent(vm *vmapi.VirtualMachine, kind vmEventKind) vmEvent {
+	info, err := api.ExtractVmInfo(vm)
+	if err != nil {
+		panic(fmt.Errorf("unexpected failure while extracting VM info from %s:%s: %w", vm.Namespace, vm.Name, err))
```

Comment:
`s/unexpected failure//`

Reply:
Resolved by making this not a panic; it shouldn't have been one.